src/Servers/Kestrel/Core/src/Internal/Http3/Http3ControlStream.cs

```diff
@@ -0,0 +1,394 @@
+                        // need to kick off httpprotocol process request async here.
+                        while (Http3FrameReader.TryReadFrame(ref readableBuffer, _incomingFrame, 16 * 1024, out var framePayload))
+                        {
+                            //Log.Http2FrameReceived(ConnectionId, _incomingFrame);
+                            consumed = examined = framePayload.End;
+                            await ProcessHttp3ControlStream(framePayload);
+                        }
+                    }
+
+                    if (result.IsCompleted)
+                    {
+                        return;
+                    }
+                }
+                catch (Http3StreamErrorException)
+                {
+                    //Log.Http2StreamError(ConnectionId, ex);
+                    //// The client doesn't know this error is coming, allow draining additional frames for now.
+                    //AbortStream(_incomingFrame.StreamId, new IOException(ex.Message, ex));
+                    //await _frameWriter.WriteRstStreamAsync(ex.StreamId, ex.ErrorCode);
+                }
+                finally
+                {
+                    Input.AdvanceTo(consumed, examined);
+
+                    //UpdateConnectionState();
+                }
+            }
+        }
+
+        private ValueTask ProcessHttp3ControlStream(in ReadOnlySequence<byte> payload)
+        {
+            // Two things:
+            // settings must be sent as the first frame of each control stream by each peer
+            // Can't send more than two settings frames.
+            switch (_incomingFrame.Type)
+            {
+                case Http3FrameType.Data:
+                case Http3FrameType.Headers:
+                case Http3FrameType.DuplicatePush:
+                case Http3FrameType.PushPromise:
+                    throw new Http3ConnectionException("HTTP_FRAME_UNEXPECTED");
+                case Http3FrameType.Settings:
+                    return ProcessSettingsFrameAsync(payload);
+                case Http3FrameType.GoAway:
+                    return ProcessGoAwayFrameAsync();
+                case Http3FrameType.CancelPush:
+                    return ProcessCancelPushFrameAsync();
+                case Http3FrameType.MaxPushId:
+                    return ProcessMaxPushIdFrameAsync();
+                default:
+                    return ProcessUnknownFrameAsync();
+            }
+        }
+
+        private ValueTask ProcessSettingsFrameAsync(ReadOnlySequence<byte> payload)
+        {
+            if (_haveReceivedSettingsFrame)
+            {
+                throw new Http3ConnectionException("H3_SETTINGS_ERROR");
+            }
+            _haveReceivedSettingsFrame = true;
+            // process a bunch of kvp of settings
+            while (true)
+            {
+                var id = VariableLengthIntegerHelper.GetVariableIntFromReadOnlySequence(payload, out var consumed, out var examinded);
```

Comment:
Why is this duplicated?

Reply:
Duplicated between what? Http3FrameReader?